test/case.lean

```diff
@@ -0,0 +1,126 @@
+import Std.Tactic.Case
+import Std.Tactic.GuardExpr
+
+set_option linter.missingDocs false
+
+example (h : x = y) (h' : z = w) : x = y ∧ z = w := by
+  constructor
+  fail_if_success show z = z
+  case _ : z = _
+  · exact h'
+  exact h
+
+example (h : x = y) (h' : z = w) : x = y ∧ z = w := by
+  constructor
+  case _ : z = _ =>
+    exact h'
+  case left =>
+    exact h
+
+example (h : x = y) (h' : z = w) : x = y ∧ z = w := by
+  constructor
+  case _ : z = _ | left => assumption
+
+example (h : x = y) (h' : z = w) : x = y ∧ z = w := by
+  constructor
+  case _ : z = _ => ?foo
+  case foo =>
+    exact h'
+  case left =>
+    exact h
+
+example (h : x = y) (h' : z = w) : x = y ∧ z + 0 = w := by
+  constructor
+  case right : z = _ =>
+    guard_target =ₛ z = w
+    exact h'
+  case _ : x = y := h
+
+example (h : x = y) : x = y ∧ x = y := by
+  constructor
+  case _ : x = y | _ : x = y => ?foo
+  -- Closes both
+  case foo => exact h
+
+example (h : x = y) : x = y ∧ x = y ∧ x = y := by
+  refine ⟨?foo, ?_, ?_⟩
+  · exact h
+  case _ : x = y | _ : x = y => ?foo
+  -- This metavariable was already assigned, so no more goals.
+
+example (h : x = y) (h' : z = w) : x = y ∧ z + 0 = w := by
+  constructor
+  case _ : z = _ =>
+    guard_target =ₛ z = w
+    exact h'
+  case left =>
+    exact h
+
+example (x y z : α) (h : x = y) (h' : y = z) : x = z := by
+  apply Eq.trans
+  case _ : α := y
+  -- Note: `case` inserts a `let_fun` due to the fact it's implemented with `show`
+  · guard_target =ₛ x = let_fun this := y; this
```

Comment:
This seems unfortunate, especially in the example given here where you wanted to fill the third goal specifically by `y`. Maybe we should just not introduce a `let_fun` if the types are reducibly defeq after unification?

Reply:
Yeah, it was unfortunate. Your idea seems reasonable, and I added a defeq detector.